jooq/src/main/java/io/micronaut/configuration/jooq/JooqConfigurationFactory.java

```diff
@@ -0,0 +1,53 @@
+package io.micronaut.configuration.jooq;
+
+import io.micronaut.context.ApplicationContext;
+import io.micronaut.context.annotation.EachBean;
+import io.micronaut.context.annotation.Factory;
+import io.micronaut.context.annotation.Parameter;
+import io.micronaut.inject.qualifiers.Qualifiers;
+import org.jooq.*;
+import org.jooq.conf.Settings;
+import org.jooq.impl.DefaultConfiguration;
+import org.jooq.impl.DefaultDSLContext;
+
+import javax.sql.DataSource;
+
+@Factory
+public class JooqConfigurationFactory {
+
+    @EachBean(DataSource.class)
+    public DefaultConfiguration jooqConfiguration(
+            @Parameter String name,
+            DataSource dataSource,
+            ApplicationContext ctx
+    ) {
+        DefaultConfiguration configuration = new DefaultConfiguration();
+
+        JooqConfigurationProperties properties = ctx.findBean(JooqConfigurationProperties.class, Qualifiers.byName(name))
+                .orElseGet(JooqConfigurationProperties::new);
+        configuration.setSQLDialect(properties.determineSqlDialect(dataSource));
+
+        configuration.setDataSource(dataSource);
+        ctx.findBean(TransactionProvider.class, Qualifiers.byName(name)).ifPresent(configuration::setTransactionProvider);
```

Comment:
All these `findBean` calls can be replaced by method parameters like `@Nullable TransactionProvider transactionProvider` then you can just check if it is `null` and avoid the manual lookups

Reply:
Will it only inject beans with the same `NameQualifier`? I vaguely remember that I got different results.

Reply:
@graemerocher I tried adding `@Nullable Settings settings` method parameter and it attempts to inject `Settings` bean ignoring any name qualifiers. I guess that's correct behaviour, otherwise it would be impossible to inject beans outside of `@EachBean` context. So I need to use `findBean` here.

Reply:
Hmmm, I just tried `@Parameter @Nullable Settings settings` and it seems to do the trick - only respectively `@Named` `Settings` bean is being injected. Is it a documented behaviour, and is it intended to work like that at all?

Reply:
It is intended to work that way, although we need to improve the documentation in this area. This is how the JPA imp works. See https://github.com/micronaut-projects/micronaut-sql/blob/master/hibernate-jpa/src/main/java/io/micronaut/configuration/hibernate/jpa/EntityManagerFactoryBean.java#L131

Reply:
I personally like compact syntax `Optional` provides here, also bean lists cannot be used with `@Parameter` so still have to use `ApplicationContext` directly. If you don't mind I will keep it like that.